agent/cache/watch.go

```diff
@@ -6,51 +6,67 @@ import (
 //
-// The returned chan is only minimally buffered so if the caller doesn't consume
-// from it quickly enough, the Watch loop will block. When the chan is later
-// drained, watching resumes correctly, however if the pause is long enough, it
-// will prevent active blocking Get on the cache and may allow the Cache-Type's
-// TTL to remove it locally. Even then when the chan is drained again, the new
-// Get will re-fetch the entry from servers and resume behaviour transparently.
+// It is a helper that abstracts code from perfroming their own "blocking" query
+// logic against a cache key to watch for changes and to maintain the key in
+// cache actively. It will continue to perform blocking Get requests until the
+// context is canceled.
 //
 // The passed context must be cancelled or timeout in order to free resources
 // and stop maintaining the value in cache. Typically request-scoped resources
 // do this but if a long-lived context like context.Background is used, then the
 // caller must arrange for it to be cancelled when the watch is no longer
 // needed.
-func (c *Cache) Watch(ctx context.Context, t string, r Request) (<-chan WatchUpdate, error) {
-	ch := make(chan WatchUpdate, 1)
+//
+// The passed chan may be buffered or unbuffered, if the caller doesn't consume
+// fast enough it will block the notification loop. When the chan is later
+// drained, watching resumes correctly. If the pause is longer than the
+// cachetype's TTL, the result might be removed from the local cache. Even in
+// this case though when the chan is drained again, the new Get will re-fetch
+// the entry from servers and resume notification behaviour transparently.
+//
+// The chan is passed in to allow multiple cached results to be watched by a
+// single consumer without juggling extra goroutines per watch. The
+// correlationID is opaque and will be returned in all UpdateEvents generated by
+// result of watching the specified request so the caller can set this to any
+// value that allows them to dissambiguate between events in the returned chan
+// when sharing a chan between multiple cache entries. If the chan is closed,
+// the notify loop will terminate.
+func (c *Cache) Notify(ctx context.Context, t string, r Request,
+	correlationID string, ch chan<- UpdateEvent) error {
 
 	// Get the type that we're fetching
 	c.typesLock.RLock()
 	tEntry, ok := c.types[t]
 	c.typesLock.RUnlock()
 	if !ok {
-		return nil, fmt.Errorf("unknown type in cache: %s", t)
+		return fmt.Errorf("unknown type in cache: %s", t)
 	}
 	if !tEntry.Type.SupportsBlocking() {
-		return nil, fmt.Errorf("watch requires the type to support blocking")
+		return fmt.Errorf("watch requires the type to support blocking")
 	}
 
 	// Always start at 0 index to deliver the inital (possibly currently cached
 	// value).
 	index := uint64(0)
 
 	go func() {
-		defer close(ch)
+		defer fmt.Println("Exited", correlationID)
```

Comment:
Probably don't want to be printing to stdout from deep in a lib. :)

Reply:
Hmmm I removed this in my last commit: https://github.com/hashicorp/consul/pull/4695/commits/c2be69354ea792beb548e5ae6793080375f5e39a

Your comment even shows up as "outdated" for me. Wonder how that happened.

Thanks though - I saw that after I pushed the change.